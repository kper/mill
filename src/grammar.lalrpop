use std::str::FromStr;
use crate::ast::*;
use lalrpop_util::ParseError;
use anyhow::Error;

grammar;

extern {
    type Error = Error;
}

Comma<T>: Vec<T> = { 
    <v:(<T> ",")*> <e:T?> => match e {
        None => v,
        Some(e) => {
            let mut v = v;
            v.push(e);
            v
        }
    }
};

SemiComma<T>: Vec<T> = { 
    <v:(<T> ";")*> => v
};

Vec<T>: Vec<T> = {
    <v:(<T>)*> => v
};

pub Pars  = Comma<Par>;

Par : Identifier = {
    <id:Id> ":" <dat:DataType> => {
        id.update_ty(dat)
    }
};
pub Program : Program = <decls:Vec<Decl>> => {
    Program {
        functions: decls.clone().into_iter()
                        .filter_map(|w| match w { Decl::Func(f) => Some(f), _ => None })
                        .collect(),
        structs: decls.clone().into_iter()
                        .filter_map(|w| match w { Decl::Struct(f) => Some(f), _ => None })
                        .collect()
    }
};

pub Decl: Decl = {
    <f:Funcdef> => Decl::Func(f),
    <s:Struct> => Decl::Struct(s)
};

pub Funcdef : Func = {
	"fn" <id:Id> "(" <pars:Pars> ")" "{" <statements:Statements> "}" =>? Func::new(id, pars, statements).map_err(|w| ParseError::User {
        error: w
    })
};

pub Struct : Struct = {
    "struct" <id:Id> "{" <fields:Comma<Field>> "}" =>? Struct::new(id, fields).map_err(|w| ParseError::User {
        error: w
    })
};

Field: Field = {
    <id:Id> ":" <dat:DataType> => Field::new(id, dat)
};

DataType: DataType = {
    "int" => DataType::Int,
    <id:Id> => DataType::Struct(Box::new(id))
}

pub Statements = SemiComma<Statement>;

pub Statement: Box<Statement> = {
	"return" <e:Expr> => Box::new(Statement::Ret(e)),
    "let" <id:Id> ":" <dat:DataType> "=" <e:Expr> => {
        Box::new(Statement::Assign(id.update_ty(dat), e))
    },
    <id:Id> ":" <cond:Conditional> => Box::new(Statement::Conditional(Some(id), cond)),
    <cond:Conditional> => Box::new(Statement::Conditional(None, cond)),
    <cond:Conditional> <id:Id> ":" => Box::new(Statement::Conditional(Some(id), cond)),
    <id:Id> "=" <e:Expr> => Box::new(Statement::ReAssign(id, e)),
    <id:Id> "." <field:Id> "=" <e:Expr> => 
        Box::new(Statement::ReAssign(id.update_field_access(Some(field)), e)),
	"let" <id:Id> "=" "new" <st:Id> =>
        Box::new(Statement::Allocate(id.update_ty(DataType::Struct(Box::new(st.clone()))), st))
};

pub Conditional: Vec<Box<Guard>> = {
    "match" <guards:SemiComma<Guard>> "end" => guards
};

pub Guard: Box<Guard> = {
    <e:Expr> "->" <stats:Statements> <cont:Continuation> => Box::new(Guard {
        guard: Some(e),
        statements: stats,
        continuation: cont,
    }),
    "_" "->" <stats:Statements> <cont:Continuation> => Box::new(Guard {
        guard: None,
        statements: stats,
        continuation: cont,
    })
};

Continuation: Continuation = {
    "continue" => Continuation::Continue,
    "break" => Continuation::Break,
};

pub Expr: Box<Expr> = {
    <n1:Term> <op:BinOp> <n2:Term> => Box::new(Expr::Dual(op, n1, n2)),
    "(" <n1:Term> <op:BinOp> <n2:Term> ")" => Box::new(Expr::Dual(Opcode::Geq, n1, n2)),
};

Exprs: Vec<Box<Expr>> = Comma<Expr>;

BinOp: Opcode = {
    "==" => Opcode::Cmp
};

pub Term: Box<Term> = {
	Num => Box::new(Term::Num(<>)),
	Id => Box::new(Term::Id(<>)),
    <obj:Id> "." <field:Id> => 
        Box::new(Term::Object(obj.update_field_access(Some(field.clone())), field)),
	"(" <n:Term> ")" => n,
	<id:Id> "(" <exprs:Exprs> ")" => Box::new(Term::Call(id, exprs)),
};

pub Id: Identifier = <l:@L> <s:r"[a-zA-Z][a-zA-Z0-9]*"> <r:@R> => Identifier::new(s.to_string(), l, r, None);
Num: i64= <s:r"[0-9]+"> => i64::from_str(s).unwrap();
