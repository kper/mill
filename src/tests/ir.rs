use crate::grammar;
use crate::tests::prelude::*;
use insta::assert_snapshot;

macro_rules! compile {
    ($input:expr) => {
        use llvm_sys::core::*;

        let _ = env_logger::builder().is_test(true).try_init();

        let input = $input;

        // Setup LLVM
        unsafe {
            let context = LLVMContextCreate();
            let module = LLVMModuleCreateWithName(c_str!("main"));
            let builder = LLVMCreateBuilderInContext(context);
            let mut codegen = Codegen::new(context, module, builder);
            let mut runner = Runner;

            // Parse
            let mut program = grammar::ProgramParser::new().parse(&input).unwrap();

            // Run the visitors
            let mut passes = crate::default_passes();
            runner
                .run_visitors(&mut passes, &mut program)
                .expect("Running visitor failed");

            // Codegen and get IR
            runner
                .run_codegen(
                    &mut CodegenVisitor::new(),
                    &mut codegen,
                    CodegenTraversal,
                    &mut program,
                )
                .expect("Codegen failed");

            let ir = crate::utils::LLVMString::new(LLVMPrintModuleToString(module)).to_string();

            assert_snapshot!(ir);
        }
    };
}

#[test]
fn test_return() {
    compile!("fn main() { return 1; }");
}

#[test]
fn test_assignment() {
    compile!("fn main() { let a : int = 1; let b : int = 2; }");
}

#[test]
fn test_reassignment() {
    compile!("fn main() { let a : int = 1; a = 2; }");
}

#[test]
fn test_addition() {
    compile!("fn main() { return 1 + 2; }");
}

#[test]
fn test_conditional() {
    compile!("fn main() -> int { if 2 == 2 { return 1; } return 0; }");
}

#[test]
fn test_conditional_with_else() {
    compile!("fn main() -> int { if 2 == 2 { return 1; } else { return 0; } return 2; }");
}

#[test]
fn test_call_when_names_in_order() {
    compile!(
        "fn f(b: int) -> int { return b; } fn main() -> int { let a : int = 1; return f(a); }"
    );
}

#[test]
fn test_call_when_names_not_in_order() {
    compile!("fn main() -> int { let a : int = 1; return f(a); } fn f(a:int) -> int { return a; }");
}

#[test]
fn test_struct() {
    compile!("struct test { a: int }");
}

#[test]
fn test_alloc_struct() {
    compile!("struct test { a: int, b: int } fn main() { let a = new test; }");
}

#[test]
fn test_field_access() {
    compile!(
        "struct t { a: int, b: int } fn main() { let o = new t; let k : int = o.a; return k; }"
    );
}

#[test]
fn test_field_assign() {
    compile!("struct t { a: int, b: int } fn main() { let o = new t; o.a = 10; let k : int = o.a; return k; }");
}
