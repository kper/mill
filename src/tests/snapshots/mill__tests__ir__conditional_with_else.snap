---
source: src/tests/ir.rs
expression: ir
---
; ModuleID = 'main'
source_filename = "main"
target triple = "x86_64-unknown-linux-gnu"

define i32 @main() {
main:
  br i1 true, label %then, label %else

then:                                             ; preds = %main
  ret i32 1
  br label %resume

else:                                             ; preds = %main
  ret i32 0
  br label %resume

resume:                                           ; preds = %else, %then
  ret i32 2
}

